### ISSUE TITLE
Verify DCPs at sync speed on low power CPU laptops

### ISSUE THREAD

**Issue by Cantar4** [wang-bin/QtAV#740]
The latest version is awesome.  With perfect color-correct images we played our mxf test-file
at around 10fps, this is remarquable in a 2012 MacBook-Air 13".  If it can reach the 24fps 
speed necessary to verify the video/audio/subtitle sync of DCP films, this app will be 
all the rage in the filmmaking community ; not only in cinema minded universities :)

To reach sync speed --without a power hungry GPU laptop-- we should transpose the trick formerly used by cinematographers to save rawstock and make it appear more sensitive : 
shot at 8fps, then show the film at 24fps by projecting each image three times ; at Aaton we 
called it [3x(1/3)].
Eight samples per second are plenty to sample natural movements to the point the trick is 
invisible to non-warned viewers. Its digital playback equivalent is utmost simple: send only 
one frame every three to the ffmeg engine (this reduces the CPU stress a lot), and once 
decoded send it three times to the ouput.

notes:
(1) this brings the possibility to play up to 30 fps 
(2) for those who own a faster CPU (like the 2016 Oct. MacBooks), a [2x(1/2)] position 
will give access to 12sps, I defy any one to see the 'subterfuge' at this sample rate.
(3) it's obvious but should be said, the subsampling doesn't apply to the audio nor 
subtitles mxf files.

**Comment by wang-bin** [comment]
Do you mean decoding 1/3 number of frames, and rendering every decoded frame 3 times to get 30 fps display rate? The motion will not be smooth, is that acceptable?

**Comment by Cantar4** [comment]
yes that is it, absolutely acceptable** and even more for a verification software on a 15" laptop.
(btw, what is the j2k tile you use ; is it 500 or 1k pixel per line ? 

**In 1969 at Eclair Corp of America we made an experiment in Los Angeles with non warned cinemagoers.  We projected at 24fps, on a commercial cinema screen, three similar sequences shot in a Tango dance school respectively shot at 24, 12 and 8sps (using a special quartz motor I designed for the EclairNPR). Nobody could see any artifacts on the 12sps shots, only a few people felt something weird on the 8sps shots. We had to show them the Oude-Delft modified 16mm projector before a second passage to make them more scrutinizing... and amazed by the elasticity of their brain ;)

**Comment by wang-bin** [comment]
And if recorded video is already at 8fps (1/3), decoded all frames and rendering at 24fps?

**Comment by wang-bin** [comment]
> btw, what is the j2k tile you use ; is it 500 or 1k pixel per line ?
What is j2k tile? Where do i find it? All I know is the mxf you sent me is 1998x1080, i.e. 1998 pixels per line

**Comment by Cantar4** [comment]
Yes it would work, but this is not the real world, you must verify the **very** DCP three hours before the public projection.  During short-film festivals I even heard stories by which the DCP monger is seated on a stool in the projection booth of a commercial cinema with an heavy laptop running easyDCP player.

**Comment by Cantar4** [comment]
[WB]_What is j2k tile? Where do i find it? All I know is the mxf you sent me is 1998x1080, i.e. 1998 pixels per line_
in the j2k encoding you find four levels of tiles each half the resolution of the upper one
this selection is done somewhere under the -lowres command, that's why I was asking !
If you are keeping our original 2K res, (which is not necessary) it means you will be mush faster at 1K (a quarter of the MB weight per frame).

**Comment by wang-bin** [comment]
What's the point of rendering at 24fps?

**Comment by wang-bin** [comment]
What's the point of rendering at 24fps?

**Comment by wang-bin** [comment]
What's the point of rendering at 24fps?

**Comment by Cantar4** [comment]
1 because 24fps is now THE smpte standard (25, 30, & 2x fps are tolerated but not all projectors can handle them).  Because with a 24fps DCP in an NTFS formatted USB key you can go in any cinema theater in the world and have your film projected in the glorious XYZ color space on a 10m wide screen in a minute. This is the beauty of DCP compared to the former 35mm film distribution
or the pixilited mp4 for TV.
This format is a must for students and independant filmmakers to be accepted by all serious film festivals.

2 because verifying a DCP must be done **at sync speed** i.e. 24fps to be sure the sound is not clipped (happens frequently with amateurish DCP makers), that the start of the audio tracks has not been shifted for any cause, that subtitles are properly 'framed' to the speech,  that there is no missing or 'green' passages.

**Comment by Cantar4** [comment]
hello wang, please be patient, I am prepping an 8sps film **_to exactly simulate_** the effect of the 3x(1/3) process out of a DCP player in verification mode.

**Comment by wang-bin** [comment]
Rendering at a specified frame rate is not hard.
On my macbook air, the frame rate is about 5fps at the original resolution. If lowres=1, sometimes it's 10~18fps and drop many frames. Sometimes it can be 24fps without dropping any frame.

**Comment by Cantar4** [comment]
@wang this is very interesting, it shows that 2x(1/2 is probably a better compromise since it allows a better quantification of the global video/audio sync.
@r-a-v-a-s  since you have access to the lowres settings in the ffmpeg guts, can you see on your retina screen the effects of the lowest resolution (btw, how many pixel per line?) on the image _"agreability"_ ?

**Comment by Cantar4** [comment]
here is the demo and how I made it : take a 24fps short sequence with white burned-in subtitles (in french in this demo) ; use a player whith a precise frame by frame mode displaying the frame count ; take screenshots carrying the # of each original image to later verify there is one every two ; batch duplicate these screenshots ; import them as free picts in DCP-o-matic ; 
add the sound track of the original sequence; add blue soft-subtitles (in english on this demo).
This is a real life DCP ready for the big screen !  but this is not the goal today,
I converted it to an email compatible 10MB MP4 for your examination.

You will see that artefacts generated by the 2x(1/2) process  are hardly visible but on Charlotte's head when she is quickly panning from left to right.  Note that her fingers on the cello cords keep all their life like vivacity, and that the slowly descending arm of Sarah keeps all his grace and smoothness.  This is more than acceptable for the verification purposes I listed yesterday :)

Btw the english soft-subtitles lost their differentiating blue color, probably due to my CLI made
 DCP to Mp4 conversion.  (there is no GUI app around to do that, weird isn't?)

**Comment by r-a-v-a-s** [comment]
@Cantar4 Sure... as soon as someone explains how I can do that... ;-]
I couldn't find any documentation for `-lowres`

**Comment by wang-bin** [comment]
@r-a-v-a-s add `av_dict_set(&d.dict, "lowres", "1", 0);` before https://github.com/wang-bin/QtAV/blob/master/src/codec/AVDecoder.cpp#L110  The 3rd parameter can be 0, 1, 2, 3, 4,...

**Comment by r-a-v-a-s** [comment]
Thanks. There doesn't seem to be much difference when testing with the mxf from Cantar4's `SarahOrigin_SHR-1_F-178_10_2K_20161003_SMPTE_OV` package.

Something I noticed in testing is that any perceived lag seems to vanish when the timeline isn't visible, regardless of the lowres mode.

**Comment by wang-bin** [comment]
@r-a-v-a-s The resolution is different, and rendering fps is higher for lowres

**Comment by Cantar4** [comment]
I just installed your nightly  build  QtAV_Player-v1.11.0-91, and yes, It plays  **realtime**  jpeg2000 cinema images extracted from a DCP on my 1.7MHz MB-Air !!!  
Done in less than three weeks, this is kind of a miracle considering  the perfectly adéquate temporal and spatial resolution  offered.    
The most difficult part being solved, remains to decode the much lighter audio.mxf  and the ultra light subtitles.mxf files to get the long sought after DCP verification player that we will carry everywhere in our small 13/15" laptops.  A huge thank to you two for this achievement
which makes QtAV the most universal player known.

**Comment by r-a-v-a-s** [comment]
> I just installed your nightly build QtAV_Player-v1.11.0-91

I sent him a build of the latest code with `av_dict_set(&d.dict, "lowres", "1", 0);` added.

**Comment by Cantar4** [comment]
I am discovering that selection of pcm_115d5c05-1083-49d9-8e17-ef719016e8bd.mxf makes us listen to the Sarah sound tracks !   I also noticed that whereas playing j2c_video.mxf occupies 100% of the MacBook CPU, playing pcm_.mxf takes 3%% only.  Which means you are very close to a full DCP player. :-]

**Comment by Cantar4** [comment]
After tests we identified two options for QtAV as a DCP previewer:

**_"lowres2 and 3x[1/3]sps"_**  will run at 24fps on a minimal laptop, i.e. CPU 1.7GHz, 4GB Ram. 
Close to DVD resolution, this will be ok for short-film verification.

**_"lowres1 and 3x[1/3]sps"_** needs a slightly more gifted machine, i.e. CPU 2.9GHz, 8GB Ram. 
It displays enjoyable real HD for feature-film verification.

Remaining tasks:
-- play  j2c_nnn.mxf
   convert XYZ to RGB, activate lowres1 image layer_____(done)
   toggle lowres2 <-> lowres1______________________(to be done)

-- play  pcm_nnn.mxf
   QtAV already plays audio files__________________________(done)

-- display  sub_nnn.mxf
   interpret mxf embeded XML subtitles_____________(semi done)
   QtAV already inserts subtitles when playing MP4s 
  
-- sync the three jobs above    ______________________(to be done)

so close to the target, we are holding our breath :)

**Comment by wang-bin** [comment]
> Displaying real HD images
What file type? I think this can be done without video player.

No problem to add lowres option in gui.

I need a video contains subtitle.

> play the three files in sync
On the same machine at the same time?

**Comment by Cantar4** [comment]
> _What file type? I think this can be done without video player._

as far as I know only your improved QtAV version in lowres-1 mode (HD equivalent)
is able to play j2k files in realtime on a mid range laptop !!!


> _No problem to add lowres option in gui._ 

good news !

> _I need a video contains subtitle._

I will send you a complete DCP containing the three files together 
I  tried to send you this package one month ago, 
we will give googledrive another chance !


> _(play the three files in sync)  On the same machine at the same time?_

Isn't what QtAV does when playing images, audio and subtitles from an MP4?

**Comment by Cantar4** [comment]
is there a command in QtAV to play images frame by frame ?

**Comment by wang-bin** [comment]
> Isn't what QtAV does when playing images, audio and subtitles from an MP4?
Usually they are in a single file. Playing external audio track (from an external audio or video file) and external subtitle file is supported.

**Comment by Cantar4** [comment]
that'is very good news !!
was the Sarah package of any help to test this function ?
If you confirm the googledrive link is not too cumbersome for yu
I will send another small DCP in which each frame carries its time code burned
in the image. 
This burned TC helps compare the performances of the CPU
when QtAV converts '12sps into 24fps' versus '8sps into 24fps'.

**Comment by wang-bin** [comment]
Does the xml subtitle have a name? Where can I find it's spec? I use ffmpeg to decode subtitles, but ffmpeg can not recognize your xml subtitles

**Comment by Cantar4** [comment]
I will send you the coordinates of Carl Etherington of DCP-o-matic fame.

**Comment by cth103** [comment]
There are two standards for DCP subtitles.  The currently widely used one is known as "Interop"; spec is here:

http://www.deluxecdn.com/dcinema/support/ti_subtitle_spec_v1_1.pdf

The newer SMPTE standard is similar, but the specification is not freely available.
